sabnzbd/cfg.py

```diff
@@ -434,6 +434,7 @@ def validate_notempty(root, value, default):
 local_ranges = OptionList("misc", "local_ranges", protect=True)
 max_url_retries = OptionNumber("misc", "max_url_retries", 10, minval=1)
 downloader_sleep_time = OptionNumber("misc", "downloader_sleep_time", 10, minval=0)
+receive_threads = OptionNumber("misc", "receive_threads", 1, minval=1)
```

Comment:
Should we not have a smarter default? Or at least 2?

Reply:
The threading adds a ton of overhead when the CPU is powerful enough to be mostly idle. For me the CPU load in Windows task manager bounces between 0.0 and 0.5% with 1 thread and the new sleep method during small downloads. With 2 threads it's between 2.5 and 4%. According to Yappi it is all threadpool stuff. I think it should only be increased from 1 if it is actually helpful for the speed.

Reply:
I'm not saying you're wrong certainly there is an overhead with threads, I just wondered if you could share any detail of what you saw in profiling to draw the conclusion that you've gone 0-0.5% to 2.5-4%, could it be attributed to being more efficient and calling the `recv_chunk` more often when using threads?

I don't really see any change in CPU usage and the cheroot webserver threadpool is taking up the most time.

Reply:
Here's a dump from Yappi after having downloaded for a while: https://gist.github.com/puzzledsab/94f137f406edf550f669a3d36a80d429

The profiling adds some overhead per function call but for most of the threading related calls on the top of the list it's a very small share and the increase in CPU usage is consistent with and without Yappi.

The new sleep time logging shows that it sleeps noticeably more frequently when using only 1 thread so it's most likely doing more downloader loops per second when not using threadpool. That's down from sleeping 0.00292 seconds about 6650 times per 20 seconds with one thread to about 6180 times with 2 threads.

Reply:
Would you mind saving a callgrind file out, `stats.save("callgrind.yappi", type="callgrind")`
I want to take a look at the callers.

I'm not sure if you can do that directly in PyCharm, the script I use is https://gist.github.com/mnightingale/f4fb48af38bed12def73ebcea679736e with the line_profiler fix I found on https://github.com/sabnzbd/sabnzbd/issues/2430#issuecomment-1414328294

Reply:
https://gist.github.com/puzzledsab/03d88c191f35cd5003509f356c890bcc

I shortened the fixed parts of the paths a bit.

Reply:
I'm not convinced that the added code-complexity is worth it if it requires manually setting a Special setting. It should work out of the box..

Reply:
> https://gist.github.com/puzzledsab/03d88c191f35cd5003509f356c890bcc
> 
> I shortened the fixed parts of the paths a bit.

Could you share one were you've set it to one thread so not using a pool?
At least 25% of the time is the webservers threads/pool.
What CPU is this running on?

I did always assume we'd have an initial based on `os.cpu_count()`.
e.g. if `os.cpu_count()` > 1 then default to 2 and probably restrict the max somewhere with `os.cpu_count()` or a lower limit.

Reply:
1 thread: https://gist.github.com/puzzledsab/e472590c73d413f13dbd76af99b76acb

@Safihre An alternative could be to do a speed test to see if it increases the speed when the number of threads is increased. In that case I think it should still be possible to set it manually in some way because in my experience the auto detection is not very reliable. I really believe it should be available considering how it can drastically increase the speed for some users. On Rpi 4 it seems to be much faster than the speeds I've seen reported for nzbget (~30 MB/s), even with AES128. I suspect it can increase the speed on other CPUs with 4 cores or more as well, although maybe not quite as much.

Reply:
I tried it on my i5, but I cannot really see a difference.. Both develop and this branch with 2 threads are almost always using 0-1% CPU. Maybe on develop it's more often 0 than 1, but it's very hard to judge.

Reply:
It's great if it's just me but I have no idea why. I'm using Windows 10 and Python 3.11.2 (just updated from 3.11.1 and 3.10 a week or so ago) from python.org. I've tried various settings for connections, downloader_sleep_time, ciphers and different servers but the difference when switching between 1 and 2 threads is always very distinct.

Reply:
you running sab on ssd or spinner? wonder if threading has more impact when you have slower disks involved or combo of higher speed + other io going on during the event

Reply:
SSD only, different disks for incomplete and complete.

Reply:
Maybe something to do with it being 4 core 4 threads? since downloader/assembler/decoder will frequently all have something to do, does your decoder have the default 2 threads?

Possibly as a default for os.cpu_count <= 4 we use **1** otherwise **2** - from my testing I never saw any meaning full improvement with more than 2.

As for keeping two paths, it's not ideal but @puzzledsab profiling does seem to show quite an impact.

Reply:
SABnzbd is about working straight out of the box, so no tweaking for regular users.
Either we default to 2 threads, or if that has a negative performance impact, we look for a different approach like the thread-per-connection.